src/main/java/edu/greenblitz/tobyDetermined/subsystems/swerve/SwerveChassis.java

```diff
@@ -0,0 +1,459 @@
+	private final Field2d field = new Field2d();
+
+	private final Ultrasonic ultrasonic;
+	private final int FILTER_BUFFER_SIZE = 15;
+	Debouncer encoderBrokenDebouncer = new Debouncer(0.2);
+	
+	
+	public static final double TRANSLATION_TOLERANCE = 0.05;
+	public static final double ROTATION_TOLERANCE = 2;
+	private boolean doVision;
+	public final double currentTolerance = 0.5;
+	
+	public SwerveChassis() {
+		this.frontLeft = new SdsSwerveModule(RobotMap.Swerve.SdsModuleFrontLeft);
+		this.frontRight = new SdsSwerveModule(RobotMap.Swerve.SdsModuleFrontRight);
+		this.backLeft = new SdsSwerveModule(RobotMap.Swerve.SdsModuleBackLeft);
+		this.backRight = new SdsSwerveModule(RobotMap.Swerve.SdsModuleBackRight);
+		this.ultrasonic = new Ultrasonic(RobotMap.Ultrasonic.PING_DIO_PORT, RobotMap.Ultrasonic.ECHO_DIO_PORT);
+		Ultrasonic.setAutomaticMode(true);
+		doVision = true;
+		
+		
+		this.pigeonGyro = new PigeonGyro(RobotMap.gyro.pigeonID);
+		
+		this.kinematics = new SwerveDriveKinematics(
+				RobotMap.Swerve.SwerveLocationsInSwerveKinematicsCoordinates
+		);
+		this.odometry = new SwerveDriveOdometry(this.kinematics, getGyroAngle(),getSwerveModulePositions());
+		this.poseEstimator = new SwerveDrivePoseEstimator(this.kinematics,
+				getPigeonAngle(),
+				getSwerveModulePositions(),
+				new Pose2d(new Translation2d(), new Rotation2d()),//Limelight.getInstance().estimateLocationByVision(),
+				new MatBuilder<>(Nat.N3(), Nat.N1()).fill(RobotMap.Vision.STANDARD_DEVIATION_ODOMETRY, RobotMap.Vision.STANDARD_DEVIATION_ODOMETRY, RobotMap.Vision.STANDARD_DEVIATION_ODOMETRY),
+				new MatBuilder<>(Nat.N3(), Nat.N1()).fill(RobotMap.Vision.STANDARD_DEVIATION_VISION2D, RobotMap.Vision.STANDARD_DEVIATION_VISION2D, RobotMap.Vision.STANDARD_DEVIATION_VISION_ANGLE));
+		SmartDashboard.putData("field", getField());
+		SmartDashboard.putData("field", getField());
+	}
+	
+	
+	public static SwerveChassis getInstance() {
+		init();
+		return instance;
+	}
+	
+	
+	public static void init() {
+		if (instance == null) {
+			instance = new SwerveChassis();
+		}
+	}
+	
+	@Override
+	public void periodic() {
+		updateOdometry();
+		updatePoseEstimationLimeLight();
+		field.setRobotPose(getRobotPose());
+	}
+	
+	public void resetAll(Pose2d pose) {
+		poseEstimator.resetPosition(getPigeonAngle(), getSwerveModulePositions(), pose);
+	}
+	
+	/**
+	 * @return returns the swerve module based on its name
+	 */
+	public SwerveModule getModule(Module module) { //TODO make private
+		switch (module) {
+			case BACK_LEFT:
+				return backLeft;
+			case BACK_RIGHT:
+				return backRight;
+			case FRONT_LEFT:
+				return frontLeft;
+			case FRONT_RIGHT:
+				return frontRight;
+		}
+		return null;
+	}
+	
+	/**
+	 * stops all the modules (power(0))
+	 */
+	public void stop() {
+		frontRight.stop();
+		frontLeft.stop();
+		backRight.stop();
+		backLeft.stop();
+	}
+	
+	/**
+	 * resetting all the angle motor's encoders to 0
+	 */
+	public void resetEncodersByCalibrationRod() {
```

Comment:
Is it still actually used?
If no, remove this.
If yes, document the meaning of the all mighty calibration rod